### PR TITLE
Enum base list completions

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Addresses feedback from https://github.com/dotnet/roslyn/pull/66119

Now completions in enum's base list include valid symbols:
devenv_qVtTFu85F0

@CyrusNajmabadi @Youssef1313 PTAL

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 1]
```diff
@@ -12024,13 +12024,343 @@ public static class Extension
...
+
+            // Not accessible in the given context
+            await VerifyItemIsAbsentAsync(source, "Byte");
+            await VerifyItemIsAbsentAsync(source, "SByte");
+            await VerifyItemIsAbsentAsync(source, "Int16");
+            await VerifyItemIsAbsentAsync(source, "UInt16");
+            await VerifyItemIsAbsentAsync(source, "Int32");
+            await VerifyItemIsAbsentAsync(source, "UInt32");
+            await VerifyItemIsAbsentAsync(source, "Int64");
+            await VerifyItemIsAbsentAsync(source, "UInt64");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:12041]: consider just defining these as the set of names we would expect, then use a theory on many of these.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 2]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
 
+        protected ImmutableArray<ISymbol> GetSymbolsForEnumBaseList(INamespaceOrTypeSymbol container = null)
+        {
+            var semanticModel = _context.SemanticModel;
+            var systemNamespace = container is not (null or INamespaceSymbol { IsGlobalNamespace: true }) ? null : (INamespaceSymbol)semanticModel.LookupNamespacesAndTypes(
+                _context.Position,
+                container: semanticModel.Compilation.GlobalNamespace,
+                name: nameof(System)).FirstOrDefault(s => s is INamespaceSymbol);
+
+            var builder = ArrayBuilder<ISymbol>.GetInstance(capacity: 9);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:293]:
```suggestion
            using var builder = ArrayBuilder<ISymbol>.GetInstance(capacity: 9);
```

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 3]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
+            AddSpecialTypeSymbolAndItsAliases(nameof(Byte), SpecialType.System_Byte);
+            AddSpecialTypeSymbolAndItsAliases(nameof(SByte), SpecialType.System_SByte);
+            AddSpecialTypeSymbolAndItsAliases(nameof(Int16), SpecialType.System_Int16);
+            AddSpecialTypeSymbolAndItsAliases(nameof(UInt16), SpecialType.System_UInt16);
+            AddSpecialTypeSymbolAndItsAliases(nameof(Int32), SpecialType.System_Int32);
+            AddSpecialTypeSymbolAndItsAliases(nameof(UInt32), SpecialType.System_UInt32);
+            AddSpecialTypeSymbolAndItsAliases(nameof(Int64), SpecialType.System_Int64);
+            AddSpecialTypeSymbolAndItsAliases(nameof(UInt64), SpecialType.System_UInt64);
+
+            return builder.ToImmutableAndFree();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:312]:
```suggestion
            return builder.ToImmutableAndClear();
```

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 4]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
+
+            void AddSpecialTypeSymbolAndItsAliases(string name, SpecialType specialType)
+            {
+                var specialTypeSymbol = _context.SemanticModel.LookupNamespacesAndTypes(
+                    _context.Position,
+                    container: container,
+                    name: name).SingleOrDefault(s => s is INamedTypeSymbol namedType && namedType.SpecialType == specialType);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:319]:
```suggestion
                var specialTypeSymbol = _context.SemanticModel
                    .LookupNamespacesAndTypes(_context.Position, container, name)
                    .SingleOrDefault(s => s is INamedTypeSymbol namedType && namedType.SpecialType == specialType);
```

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 5]
```diff
@@ -111,6 +111,10 @@ private ImmutableArray<ISymbol> GetSymbolsForCurrentContext()
             {
                 return GetSymbolsForNamespaceDeclarationNameContext<BaseNamespaceDeclarationSyntax>();
             }
+            else if (_context.IsEnumBaseListContext)
+            {
+                return GetSymbolsForEnumBaseList();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:116]: i'd prefer you to explicitly pass `container: null` here.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 6]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
             return symbols;
         }
 
+        protected ImmutableArray<ISymbol> GetSymbolsForEnumBaseList(INamespaceOrTypeSymbol container = null)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:285]: make non-optional plz :)

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 7]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
+        {
+            var semanticModel = _context.SemanticModel;
+            var systemNamespace = container is not (null or INamespaceSymbol { IsGlobalNamespace: true }) ? null : (INamespaceSymbol)semanticModel.LookupNamespacesAndTypes(
+                _context.Position,
+                container: semanticModel.Compilation.GlobalNamespace,
+                name: nameof(System)).FirstOrDefault(s => s is INamespaceSymbol);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:291]:
```suggestion
            var systemNamespace = container is not (null or INamespaceSymbol { IsGlobalNamespace: true })
                ? null
                 : semanticModel.LookupNamespacesAndTypes(_context.Position, semanticModel.Compilation.GlobalNamespace, nameof(System))
                     .OfType<INamespaceSymbol>().FirstOrDefault();
```

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 8]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
+
+            var builder = ArrayBuilder<ISymbol>.GetInstance(capacity: 9);
+
+            if (systemNamespace is not null)
+            {
+                builder.Add(systemNamespace);
+
+                var aliases = semanticModel.LookupSymbols(_context.Position, container).Where(s => s is IAliasSymbol alias && alias.Target.Equals(systemNamespace));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:299]:
```suggestion
                var aliases = semanticModel.LookupSymbols(_context.Position, container).OfType<IAliasSymbol>().Where(a => systemNamespace.Equals(a.Target));
```

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 9]
```diff
@@ -282,6 +282,51 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
+                var specialTypeSymbol = _context.SemanticModel.LookupNamespacesAndTypes(
+                    _context.Position,
+                    container: container,
+                    name: name).SingleOrDefault(s => s is INamedTypeSymbol namedType && namedType.SpecialType == specialType);
+
+                builder.AddIfNotNull(specialTypeSymbol);
+
+                specialTypeSymbol ??= _context.SemanticModel.Compilation.GetSpecialType(specialType);
+
+                var aliases = _context.SemanticModel.LookupSymbols(_context.Position, container).Where(s => s is IAliasSymbol alias && alias.Target.Equals(specialTypeSymbol));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:325]:
```suggestion
                var aliases = _context.SemanticModel.LookupSymbols(_context.Position, container).OfType<IAliasSymbol>().Where(a => specialTypeSymbol.Equals(s.Target));
```

**Comment by CyrusNajmabadi** [comment]
Looking good :)

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 10]
```diff
@@ -12023,14 +12023,251 @@ public static class Extension
             await VerifyItemIsAbsentAsync(MakeMarkup(source), "ExtMethod");
         }
 
-        [Fact]
-        public async Task NoSymbolCompletionsInEnumBaseList()
+        [Theory]
+        [MemberData(nameof(ValidEnumUnderlyingTypeNames))]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:12027]:
```suggestion
        [Theory, MemberData(nameof(ValidEnumUnderlyingTypeNames))]
```

and the rest.

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb [hunk 11]
```diff
@@ -8304,5 +8304,322 @@ End Class"
...
+
+            ' Not accessible in the given context
+            Await VerifyItemIsAbsentAsync(markup, "Byte")
+            Await VerifyItemIsAbsentAsync(markup, "SByte")
+            Await VerifyItemIsAbsentAsync(markup, "Int16")
+            Await VerifyItemIsAbsentAsync(markup, "UInt16")
+            Await VerifyItemIsAbsentAsync(markup, "Int32")
+            Await VerifyItemIsAbsentAsync(markup, "UInt32")
+            Await VerifyItemIsAbsentAsync(markup, "Int64")
+            Await VerifyItemIsAbsentAsync(markup, "UInt64")
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb:8322]: theory as well.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:285]: still applies.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 12]
```diff
@@ -282,6 +282,50 @@ protected ImmutableArray<ISymbol> GetSymbolsForNamespaceDeclarationNameContext<T
...
+            AddSpecialTypeSymbolAndItsAliases(nameof(Int64), SpecialType.System_Int64);
+            AddSpecialTypeSymbolAndItsAliases(nameof(UInt64), SpecialType.System_UInt64);
+
+            return builder.ToImmutableAndClear();
+
+            void AddSpecialTypeSymbolAndItsAliases(string name, SpecialType specialType)
+            {
+                var specialTypeSymbol = _context.SemanticModel
+                    .LookupNamespacesAndTypes(_context.Position, container, name)
+                    .SingleOrDefault(s => s is INamedTypeSymbol namedType && namedType.SpecialType == specialType);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs:318]:
```suggestion
                    .FirstOrDefault(s => s is INamedTypeSymbol namedType && namedType.SpecialType == specialType);
```

.Single can throw.  Unlikely, but i don't ever want to have to deal with a bug on that.

**Comment by DoctorKrolic** [comment]
@CyrusNajmabadi CI is ok. Merge?

**Comment by CyrusNajmabadi** [comment]
thanks!

**Pull request merged**